django/forms/widgets.py

```diff
@@ -115,39 +117,41 @@ def __getitem__(self, name):
         raise KeyError('Unknown media type "%s"' % name)
 
     @staticmethod
-    def merge(list_1, list_2):
+    def merge(*lists):
         """
-        Merge two lists while trying to keep the relative order of the elements.
-        Warn if the lists have the same two elements in a different relative
-        order.
+        Merge several lists while trying to keep the relative order of the elements.
+        Warn if the lists have the same elements in a different relative order.
 
         For static assets it can be important to have them included in the DOM
         in a certain order. In JavaScript you may not be able to reference a
         global or in CSS you might want to override a style.
         """
-        # Start with a copy of list_1.
-        combined_list = list(list_1)
-        last_insert_index = len(list_1)
-        # Walk list_2 in reverse, inserting each element into combined_list if
-        # it doesn't already exist.
-        for path in reversed(list_2):
-            try:
-                # Does path already exist in the list?
-                index = combined_list.index(path)
-            except ValueError:
-                # Add path to combined_list since it doesn't exist.
-                combined_list.insert(last_insert_index, path)
-            else:
-                if index > last_insert_index:
-                    warnings.warn(
-                        'Detected duplicate Media files in an opposite order:\n'
-                        '%s\n%s' % (combined_list[last_insert_index], combined_list[index]),
-                        MediaOrderConflictWarning,
-                    )
-                # path already exists in the list. Update last_insert_index so
-                # that the following elements are inserted in front of this one.
-                last_insert_index = index
-        return combined_list
+
+        dependency_graph = defaultdict(set)
+        unique_items = []
+
+        for lst in filter(None, lists):
+            head = lst[0]
+            # First entries are always part of the dependency graph.
+            dependency_graph.setdefault(head, set())
+            for item in lst:
+                if item not in unique_items:
```

Comment:
I cannot use a set since the order of `unique_items` is significant. `unique_items` defines the ordering of assets of the same dependency depth and also acts as a fallback when we have circular dependencies which cannot be resolved using the topological sort.

I could use a `dict` or an `OrderedDict` and only use the keys later but I worry a bit that's not seem very obvious to me.

Reply:
It's probably fine https://github.com/matthiask/django/commit/993758e1f9eb96e231162a1c3c2dabc97db7fbd2

If this change is backported to a Django version supporting Python 3.5 the dict has to be changed to an `OrderedDict`.

Reply:
Resolved by using `django.utils.datastructures.OrderedSet` (I didn't know this existed)